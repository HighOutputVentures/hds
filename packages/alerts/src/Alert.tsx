import {
  Alert as ChakraAlert,
  AlertDescription,
  AlertIcon,
  AlertTitle,
  chakra,
  Fade,
  Icon,
  SystemStyleObject,
} from '@chakra-ui/react';
import { ComponentProps, ReactNode } from 'react';
import CheckCircleIcon from './CheckCircleIcon';
import CloseIcon from './CloseIcon';
import ExclamationCircle from './ExclamationCircle';
import ExclamationTriangleIcon from './ExclamationTriangle';
import InformationCircleIcon from './InformationCircle';

type Accent = 'primary' | 'warning' | 'success' | 'error' | 'gray';

export interface AlertProps {
  isOpen?: boolean;
  accent?: Accent;
  title: ReactNode;
  description: ReactNode;
  primaryButton?: string | boolean;
  secondaryButton?: string | boolean;
  onPrimaryButtonClick?(): void;
  onSecondaryButtonClick?(): void;
  onClose?(): void;

  __testId?: string;
  __iconTestId?: string;
  __titleTestId?: string;
  __descriptionTestId?: string;
  __closeBtnTestId?: string;
  __primaryBtnTestId?: string;
  __secondaryBtnTestId?: string;
}

export function Alert(props: AlertProps & SystemStyleObject) {
  const {
    accent = 'primary',
    isOpen = true,
    title,
    description,
    primaryButton,
    secondaryButton,
    onPrimaryButtonClick,
    onSecondaryButtonClick,
    onClose,

    __testId = 'hds.alert',
    __iconTestId = 'hds.alert.icon',
    __titleTestId = 'hds.alert.title',
    __descriptionTestId = 'hds.alert.description',
    __closeBtnTestId = 'hds.alert.close-btn',
    __primaryBtnTestId = 'hds.alert.primary-btn',
    __secondaryBtnTestId = 'hds.alert.secondary-btn',

    ...others
  } = props;

  const SVGIcon = (props: ComponentProps<'svg'>) => {
    switch (accent) {
      case 'warning':
        return <ExclamationTriangleIcon {...props} />;
      case 'error':
        return <ExclamationCircle {...props} />;
      case 'success':
        return <CheckCircleIcon {...props} />;
      default:
        return <InformationCircleIcon {...props} />;
    }
  };

  return (
    <Fade in={isOpen}>
      <ChakraAlert
        variant="unstyled"
        sx={{
          gap: '12px',
          padding: '16px',
          display: 'flex',
          alignItems: 'start',
          rounded: '8px',
          border: '1px solid',

          ...(accent === 'primary' && {
            bgColor: 'brand.primary.500',
            borderColor: 'brand.primary.600',
          }),

          ...(accent === 'gray' && {
            bgColor: 'Gray.25',
            borderColor: 'Gray.300',
          }),

          ...(accent === 'error' && {
            bgColor: 'interface.error.500',
            borderColor: 'interface.error.600',
          }),

          ...(accent === 'warning' && {
            bgColor: 'interface.warning.500',
            borderColor: 'interface.warning.600',
          }),

          ...(accent === 'success' && {
            bgColor: 'interface.success.500',
            borderColor: 'interface.success.600',
          }),

          ...others,
        }}
        data-testid={__testId}
      >
        <AlertIcon
          as={SVGIcon}
          sx={{
            width: '20px',
            height: '20px',

            ...(accent === 'gray' && { color: 'neutrals.800' }),
            ...(accent === 'error' && { color: 'interface.error.800' }),
            ...(accent === 'warning' && { color: 'interface.warning.800' }),
            ...(accent === 'success' && { color: 'interface.success.800' }),
            ...(accent === 'primary' && { color: 'brand.primary.900' }),
          }}
          data-testid={__iconTestId}
        />

        <chakra.div flexGrow={1}>
          <AlertTitle
            sx={{
              fontSize: '14px',
              fontWeight: 'medium',
              lineHeight: '20px',

              ...(accent === 'gray' && { color: 'neutrals.900' }),
              ...(accent === 'error' && { color: 'interface.error.900' }),
              ...(accent === 'warning' && { color: 'interface.warning.900' }),
              ...(accent === 'success' && { color: 'interface.success.900' }),
              ...(accent === 'primary' && { color: 'brand.primary.900' }),
            }}
            data-testid={__titleTestId}
          >
            {title}
          </AlertTitle>
          <AlertDescription
            sx={{
              marginTop: '4px',
              fontSize: '14px',
              lineHeight: '20px',

              ...(accent === 'gray' && { color: 'neutrals.600' }),
              ...(accent === 'error' && { color: 'interface.error.700' }),
              ...(accent === 'success' && { color: 'interface.success.800' }),
              ...(accent === 'warning' && { color: 'interface.warning.800' }),
              ...(accent === 'primary' && { color: 'brand.primary.700' }),
            }}
            data-testid={__descriptionTestId}
          >
            {description}
          </AlertDescription>

          <chakra.div
            sx={{
              marginTop: '12px',
              display: 'flex',
              gap: '12px',
            }}
          >
            {!!secondaryButton && (
              <chakra.button
                onClick={onSecondaryButtonClick}
                sx={{
                  fontWeight: 500,
                  fontSize: '16px',
                  lineHeight: '20px',

                  ...(accent === 'gray' && { color: 'neutrals.700' }),
                  ...(accent === 'error' && { color: 'interface.error.700' }),
                  ...(accent === 'warning' && { color: 'interface.warning.800' }),
                  ...(accent === 'success' && { color: 'interface.success.800' }),
                  ...(accent === 'primary' && { color: 'brand.primary.800' }),
                }}
                data-testid={__secondaryBtnTestId}
              >
                {secondaryButton}
              </chakra.button>
            )}

            {!!primaryButton && (
              <chakra.button
                onClick={onPrimaryButtonClick}
                sx={{
                  fontWeight: 500,
                  fontSize: '16px',
                  lineHeight: '20px',

                  ...(accent === 'gray' && { color: 'brand.primary.700' }),
                  ...(accent === 'error' && { color: 'interface.error.900' }),
                  ...(accent === 'warning' && { color: 'interface.warning.900' }),
                  ...(accent === 'success' && { color: 'interface.success.900' }),
                  ...(accent === 'primary' && { color: 'brand.primary.900' }),
                }}
                data-testid={__primaryBtnTestId}
              >
                {primaryButton}
              </chakra.button>
            )}
          </chakra.div>
        </chakra.div>

        <chakra.div>
          <chakra.button
            onClick={onClose}
            sx={{
              ...(accent === 'gray' && { color: 'brand.primary.700' }),
              ...(accent === 'error' && { color: 'interface.error.700' }),
              ...(accent === 'success' && { color: 'interface.success.700' }),
              ...(accent === 'warning' && { color: 'interface.warning.700' }),
              ...(accent === 'primary' && { color: 'brand.primary.700' }),
            }}
            data-testid={__closeBtnTestId}
          >
            <Icon as={CloseIcon} w="20px" h="20px" />
          </chakra.button>
        </chakra.div>
      </ChakraAlert>
    </Fade>
  );
}
