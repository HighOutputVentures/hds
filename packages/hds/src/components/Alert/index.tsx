export {
  Alert,
  AlertDescription,
  AlertDescriptionProps,
  AlertDialog,
  AlertDialogBody,
  AlertDialogCloseButton,
  AlertDialogContent,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogOverlay,
  AlertDialogProps,
  AlertIcon,
  AlertIconProps,
  AlertProps,
  AlertTitle,
  AlertTitleProps,
} from '@chakra-ui/react';
