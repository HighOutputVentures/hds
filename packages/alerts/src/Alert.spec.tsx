import { ChakraProvider } from '@chakra-ui/react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { Alert } from './Alert';

const onClose = jest.fn();
const onPrimaryButtonClick = jest.fn();
const onSecondaryButtonClick = jest.fn();

describe('Alert', () => {
  beforeEach(() => {
    render(
      <ChakraProvider>
        <Alert
          primaryButton="View Changes"
          secondaryButton="Learn More"
          accent="primary"
          title="We&rsquo;ve just released a new feature"
          description="Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam."
          onClose={onClose}
          onPrimaryButtonClick={onPrimaryButtonClick}
          onSecondaryButtonClick={onSecondaryButtonClick}
        />
      </ChakraProvider>,
    );
  });

  it('Should render', () => {
    expect(screen.getByTestId('hds.alert')).toBeInTheDocument();
  });

  it('should render title', () => {
    expect(screen.getByTestId('hds.alert.title')).toBeInTheDocument();
  });

  it('should render description', () => {
    expect(screen.getByTestId('hds.alert.description')).toBeInTheDocument();
  });

  it('should render primary button', () => {
    expect(screen.getByTestId('hds.alert.primary-btn')).toBeInTheDocument();
  });

  it('should render secondary button', () => {
    expect(screen.getByTestId('hds.alert.secondary-btn')).toBeInTheDocument();
  });

  it('should render close button', () => {
    expect(screen.getByTestId('hds.alert.close-btn')).toBeInTheDocument();
  });

  test("'onClose'", async () => {
    await userEvent.click(screen.getByTestId('hds.alert.close-btn'));
    expect(onClose).toHaveBeenCalled();
  });

  test("'onPrimaryButtonClick'", async () => {
    await userEvent.click(screen.getByTestId('hds.alert.primary-btn'));
    expect(onPrimaryButtonClick).toHaveBeenCalled();
  });

  test("'onSecondaryButtonClick'", async () => {
    await userEvent.click(screen.getByTestId('hds.alert.secondary-btn'));
    expect(onSecondaryButtonClick).toHaveBeenCalled();
  });

  describe('Snapshot', () => {
    it('Should match snapshot', () => {
      const component = renderer.create(
        <Alert
          primaryButton="View Changes"
          secondaryButton="Learn More"
          accent="primary"
          title="We&rsquo;ve just released a new feature"
          description="Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam."
        />,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      component.unmount();
    });
  });
});
