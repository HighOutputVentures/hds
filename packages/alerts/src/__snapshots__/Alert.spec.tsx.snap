// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Snapshot Should match snapshot 1`] = `
<div
  className="chakra-fade"
  style={
    {
      "opacity": 0,
    }
  }
>
  <div
    className="chakra-alert css-f4qb0s"
    data-status="info"
    data-testid="hds.alert"
    role="alert"
  >
    <svg
      className="chakra-alert__icon css-1g1jahg"
      data-status="info"
      data-testid="hds.alert.icon"
      fill="none"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clipPath="url(#clip0_1130_82786)"
      >
        <path
          d="M10.0001 13.3334V10M10.0001 6.66669H10.0084M18.3334 10C18.3334 14.6024 14.6025 18.3334 10.0001 18.3334C5.39771 18.3334 1.66675 14.6024 1.66675 10C1.66675 5.39765 5.39771 1.66669 10.0001 1.66669C14.6025 1.66669 18.3334 5.39765 18.3334 10Z"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="1.66667"
        />
      </g>
      <defs>
        <clipPath
          id="clip0_1130_82786"
        >
          <rect
            fill="white"
            height="20"
            width="20"
          />
        </clipPath>
      </defs>
    </svg>
    <div
      className="css-i9gxme"
    >
      <div
        className="chakra-alert__title css-v7ucr"
        data-status="info"
        data-testid="hds.alert.title"
      >
        Weâ€™ve just released a new feature
      </div>
      <div
        className="chakra-alert__desc css-1t3j939"
        data-status="info"
        data-testid="hds.alert.description"
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Aliquid pariatur, ipsum similique veniam.
      </div>
      <div
        className="css-13aw5n7"
      >
        <button
          className="css-49mbms"
          data-testid="hds.alert.secondary-btn"
        >
          Learn More
        </button>
        <button
          className="css-1sdfhj1"
          data-testid="hds.alert.primary-btn"
        >
          View Changes
        </button>
      </div>
    </div>
    <div
      className="css-0"
    >
      <button
        className="css-14sm3cc"
        data-testid="hds.alert.close-btn"
      >
        <svg
          className="chakra-icon css-gkivye"
          fill="none"
          focusable={false}
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 5L5 15M5 5L15 15"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="1.66667"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
